modules/api/src/main/scala/cards/nine/api/NineCardsDirectives.scala

```diff
@@ -54,46 +53,59 @@ class NineCardsDirectives(
   with PathDirectives
   with RouteDirectives
   with SecurityDirectives
+  with FutureDirectives
   with JsonFormats {
 
   import cards.nine.api.accounts.JsonFormats._
 
-  implicit def fromTaskAuth[T](auth: ⇒ Task[Authentication[T]]): AuthMagnet[T] =
-    new AuthMagnet(onSuccess(auth))
+  implicit def fromTaskAuth[T](auth: ⇒ Task[AuthenticationResult[T]]): Future[AuthenticationResult[T]] =
+    auth.unsafePerformAsyncFuture()
+
+  val nineCardChallenge = HttpChallenge("NineCard", None)
 
   val rejectionByCredentialsRejected = AuthenticationFailedRejection(
-    cause            = AuthenticationFailedRejection.CredentialsRejected,
-    challengeHeaders = Nil
+    cause     = AuthenticationFailedRejection.CredentialsRejected,
+    challenge = nineCardChallenge
   )
 
+  def authenticate[A](result: Future[AuthenticationResult[A]]): Directive1[A] =
+    onSuccess(result).flatMap {
+      case Right(a) ⇒ provide(a)
+      case Left(b) ⇒ reject(rejectionByCredentialsRejected)
+    }
+
   val authenticateLoginRequest: Directive1[SessionToken] = for {
     request ← entity(as[ApiLoginRequest])
     _ ← authenticate(validateLoginRequest(request.email, request.tokenId))
     sessionToken ← generateSessionToken
   } yield sessionToken
 
-  def validateLoginRequest(email: Email, tokenId: GoogleIdToken): Task[Authentication[Unit]] =
+  def validateLoginRequest(email: Email, tokenId: GoogleIdToken): Task[AuthenticationResult[Unit]] =
     if (email.value.isEmpty || tokenId.value.isEmpty)
-      Task.now(Left(rejectionByCredentialsRejected))
+      Task.now(Left(nineCardChallenge))
     else
       accountProcesses
         .checkGoogleTokenId(email, tokenId)
-        .leftMap(_ ⇒ rejectionByCredentialsRejected)
+        .leftMap(_ ⇒ nineCardChallenge)
         .value
         .foldMap(prodInterpreters)
         .handle {
-          case _ ⇒ Left(rejectionByCredentialsRejected)
+          case _ ⇒ Left(nineCardChallenge)
         }
 
-  val authenticateUser: Directive1[UserContext] = for {
-    uri ← requestUri
-    herokuForwardedProtocol ← optionalHeaderValueByName(headerHerokuForwardedProto)
-    herokuUri = herokuForwardedProtocol.filterNot(_.isEmpty).fold(uri)(uri.withScheme)
-    sessionToken ← headerValueByName(headerSessionToken).map(SessionToken)
-    androidId ← headerValueByName(headerAndroidId).map(AndroidId)
-    authToken ← headerValueByName(headerAuthToken)
-    userId ← authenticate(validateUser(sessionToken, androidId, authToken, herokuUri))
-  } yield UserContext(UserId(userId), androidId) :: HNil
+  // TODO get back to the single for comprehension
+  val authenticateUser: Directive1[UserContext] =
+    extractUri.flatMap { uri ⇒
+      optionalHeaderValueByName(headerHerokuForwardedProto).flatMap { herokuForwardedProtocol ⇒
+        val herokuUri = herokuForwardedProtocol.filterNot(_.isEmpty).fold(uri)(uri.withScheme)
```

Comment:
In akka-http, `Directive1#flatMap` is provided via [an implicit class](https://github.com/akka/akka-http/blob/v10.0.4/akka-http/src/main/scala/akka/http/scaladsl/server/Directive.scala#L136-L149) and `herokuUri` breaks the for comprehension for a strange reason. Need to revisit in Scala 2.12